File: #1079.letter-tile-possibilities.cpp
class Solution {
public:
    int numTilePossibilities(string tiles) {

        unordered_map<char, int> tile;
        vector<int> num;
        for (auto &c : tiles) {
            if (tile .find(c) == tile.end()) {
                tile[c] = num.size();
                num.push_back(0);
            }
            num[tile[c]]++;
        }
        result = 0;
        numTilePossibilities(num);
        return result - 1;
    }

private:
    int result;

    void numTilePossibilities(vector<int> &num) {

        result++;
        for (int i = 0; i < num.size(); i++) {
            if (num[i] > 0) {
                num[i]--;
                numTilePossibilities(num);
                num[i]++;
            }
        }
    }
};

File: #1079.letter-tile-possibilities.js
/**
 * @param {string} tiles
 * @return {number}
 */
var numTilePossibilities = function(tiles) {
    return search(tiles);
};

const isValidState = (state = "", res=[]) => !!state && !res.includes(state);
const search = (tiles = "", state = '', result=[]) => {
    if (isValidState(state,result)) {
        result.push(state);
    }
   
    for (let i = 0; i < tiles.length; i++) {
        state += tiles[i];
        const newTiles = tiles.substring(0, i) + tiles.substring(i + 1);
        search(newTiles, state,result);
        state = state.slice(0, state.length - 1);
    }
    return result.length;
}
